Fix sa_add_projid to lookup and update SA_ZPL_DXATTR (avoid DXATTR loss)

jsai20 opened the pull request:
sa_add_projid() gets called via zfs_setattr() for setting project id on old file/dir, which were created before upgrading to project quota feature. This function does lookup for all possible SA and update them all together along with project ID at needed fixed offset. But its missing lookup and update of SA_ZPL_DXATTR, effectively it losses SA_ZPL_DXATTR.

Signed-off-by: Jitendra Patidar <jitendra.patidar@nutanix.com>
Closes #16287

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

jsai20 commented:
[project-quota-dxattr-test-log.txt](https://github.com/user-attachments/files/15930186/project-quota-dxattr-test-log.txt)

tonyhutter commented:
I'm new to this particular bit of code and an trying to understand what is being fixed.  Is it that when you update old, pre-project-quota files with `sa_add_projid()`, that the project ID is not getting set in the SA xattrs, and this PR fixes it?

jsai20 commented:
> @jsai20 I'm new to this particular bit of code and an trying to understand what is being fixed. Is it that when you update old, pre-project-quota files with `sa_add_projid()`, that the project ID is not getting set in the SA xattrs, and this PR fixes it?

@tonyhutter
 Following header comment on sa_add_projid(), explains why sa_add_projid() needed.
/*
 * For the existed object that is upgraded from old system, its ondisk layout
 * has no slot for the project ID attribute. But quota accounting logic needs
 * to access related slots by offset directly. So we need to adjust these old
 * objects' layout to make the project ID to some unified and fixed offset.
 */
#define SA_PROJID_OFFSET        128

So, sa_add_projid() function basically expected to do lookup for all possible existing SA on object, reorder them
to fix offset for project id. And finally update all SA's together. 

sa_add_projid() currently doesn't lookup and update of SA_ZPL_DXATTR, effectively it losses
existing SA_ZPL_DXATTR on object.

tonyhutter commented:
Ok, so it basically allows you to set project ID on a upgraded pool with xattr=sa.

tonyhutter commented:
Looking for one more reviewer on this

robn commented on module/zfs/sa.c:
```diff
@@ -2237,6 +2277,7 @@ EXPORT_SYMBOL(sa_bulk_lookup);
 EXPORT_SYMBOL(sa_bulk_lookup_locked);
 EXPORT_SYMBOL(sa_bulk_update);
 EXPORT_SYMBOL(sa_size);
+EXPORT_SYMBOL(sa_size_locked);
```
Is there a reason to export it? Or just for completeness?

(I don't think we really have any policy for what we export?)

jsai20 replied:
Not needed. Removed it.

robn reviewed (approved):
I don't know anything about DXATTR, but the logic seems right when compared to the way it's used elsewhere.

A couple of structure nits, but I won't hold it up because of them. Good work, thanks!

jsai20 commented on module/zfs/sa.c:
```diff
@@ -2063,6 +2083,26 @@ sa_update(sa_handle_t *hdl, sa_attr_type_t type,
+}
+
 int
 sa_size(sa_handle_t *hdl, sa_attr_type_t attr, int *size)
```
done

jsai20 commented:
> I don't know anything about DXATTR, but the logic seems right when compared to the way it's used elsewhere.
> 
> A couple of structure nits, but I won't hold it up because of them. Good work, thanks!

Thanks @robn. I just updated a new patch addressing your comments.

robn commented:
LGTM!
